tarantool/dbapi.py

```diff
@@ -0,0 +1,227 @@
+    def lastrowid(self):
+        """
+        This read-only attribute provides the rowid of the last modified row
+        (most databases return a rowid only when a single INSERT operation is
+        performed).
+        """
+        return self._lastrowid
+
+    @property
+    def rowcount(self):
+        """
+        This read-only attribute specifies the number of rows that the last
+        .execute*() produced (for DQL statements like SELECT) or affected (
+        for DML statements like UPDATE or INSERT).
+        """
+        return self._rowcount
+
+    def _check_result_set(self, error=None):
+        """
+        Closed method for raising an error when Cursor object does not have
+        any row to fetch. Useful for checking access after DQL requests.
+        """
+        if self._rows is None:
+            raise ProgrammingError(error or "Result set is empty.")
+
+    def fetchone(self):
+        """
+        Fetch the next row of a query result set, returning a single
+        sequence, or None when no more data is available.
+        """
+        self._check_result_set()
+        return self.fetchmany(1)[0] if self._rows else None
+
+    def fetchmany(self, size=None):
+        """
+        Fetch the next set of rows of a query result, returning a sequence of
+        sequences (e.g. a list of tuples). An empty sequence is returned when
+        no more rows are available.
+        """
+        self._check_result_set()
+
+        size = size or self.arraysize
+
+        if len(self._rows) < size:
+            items = self._rows
+            self._rows = []
+        else:
+            items, self._rows = self._rows[:size], self._rows[size:]
+
+        return items
+
+    def fetchall(self):
+        """Fetch all (remaining) rows of a query result, returning them as a
+        sequence of sequences (e.g. a list of tuples). Note that the cursor's
+        arraysize attribute can affect the performance of this operation.
+        """
+        self._check_result_set()
+
+        items = self._rows
+        self._rows = []
+        return items
+
+    def setinputsizes(self, sizes):
+        """PEP-249 allows to not implement this method and do nothing."""
+
+    def setoutputsize(self, size, column=None):
+        """PEP-249 allows to not implement this method and do nothing."""
+
+
+class Connection(BaseConnection):
+
+    @property
+    def server_version(self):
+        """
+        Returns the server version from greeting
+        """
+        return getattr(self, 'version_id')
+
+    def _check_not_closed(self, error=None):
+        """
+        Checks if the connection is not closed and rises an error if it is.
+        """
+        if self._socket is None:
+            raise DatabaseError(error or "The connector is closed")
+
+    def close(self):
+        """
+        Closes the connection
+        """
+        self._check_not_closed("The closed connector can not be closed again.")
+        super(Connection, self).close()
```

Comment:
Aside of this, it [must](https://www.python.org/dev/peps/pep-0249/#connection-close) close all its cursors:

> The connection will be unusable from this point forward; an Error (or subclass) exception will be raised if any operation is attempted with the connection. The same applies to all cursor objects trying to use the connection.

Reply:
I have added an extra check to _check_not_closed() Cursor method.
```
    def _check_not_closed(self, error=None):
        if self._c is None:
            raise ProgrammingError(error or "Can not operate on closed cursor")
        if self._c.is_closed():
            raise ProgrammingError("The cursor can not be used "
                                   "with a closed connection")
```
Do you think it is enough for closing Cursor after connection was closed?

Reply:
Should work. I'll look thoroughly at the next review iteration.